### PR TITLE
add option to not HUP on reload

### LINKED ISSUE #22: Make HUP optional for reloading
Reload currently performs a HUP every time a configuration is set, it would be nice to have an option to not do that if Faucet is using the environment variable `FAUCET_CONFIG_STAT_RELOAD=1` which automatically reloads the config whenever it detects the config has changed.

### PR DESCRIPTION
**Pull request by cglewis**
This closes #22.

It needs to wait for #25 to be merged first for the linting to pass.

### CODE REVIEW

**Comment by lantz** [comment]
Unfortunately it looks like FAUCET doesn't update `config_hash_info` in this situation, so we'll have to fix that first.

**Comment by cglewis** [comment]
Ok, those tests are not working as I expected. I'll have to mess with this some more.

**Comment by lantz** [comment]
Nice - this is getting there; I think that `FAUCET_CONFIG_STAT_RELOAD` should be set appropriately in the `FAUCET` controller class. Usually the idiom is to add an initialization parameter to the controller constructor, and then pass it into the network constructor, something like:

```python
faucet = partial(FAUCET, config_stat_reload=1)
net = Mininet(topo=TestTopo(), controller=faucet, autoSetMacs=True)
```

I think something like this will be good on the agent side; then we'll need to change FAUCET to make sure that `faucet_config_hash_info` is updated properly when a reload is triggered due to `FAUCET_CONFIG_STAT_RELOAD`.

**Comment by cglewis** [comment]
Ok, I think this is closer, though it's still not applying the environment variable to the `config_stat_reload` so something is not quite right. There's also something still causing the `config_files` error in the tests that I haven't tracked down yet. I'll have to get back to this a bit later.

**Comment by cglewis** [comment]
tests for this should pass now once https://github.com/faucetsdn/faucet/pull/2979 is merged in. I had to remove the writing an empty config file, as Faucet never gets out of the error state when using `CONFIG_STAT_RELOAD`. I'm not entirely sure why that behavior is different then HUP, but since it's an error that's getting thrown in both cases, it seemed like it was cleaner to remove anyway.

**Comment by lantz** [comment]
If the FAUCET patch generates the prometheus var when a config load fails does this mean the original agent test will pass?

**Comment by cglewis** [comment]
@lantz yes, that is my understanding.

**Comment by cglewis** [comment]
@lantz is this good to go, or does it need anything else before it can be merged?

File: faucetagent.py [hunk 1]
```diff
@@ -339,6 +342,10 @@ def parse():
         default=0.0,
         help='Wait for FAUCET to attempt to update this fraction of DPs')
     arg('--timeout', default=120, help='max time(s) to wait for config reload')
+    arg('--nohup',
+        type=int,
```
Comment (lantz) [inline faucetagent.py]: Should probably be something like `arg('--nohup', action='store_true')`

File: agenttest.py [hunk 2]
```diff
@@ -332,21 +332,20 @@ def send_arps(hosts):
 # pylint: disable=too-many-locals, too-many-statements
 
 
-def end_to_end_test():
+def end_to_end_test(options, envs):
```
Comment (lantz) [inline agenttest.py]: The right thing to do here is to pass in the arguments as parameters with defaults, e.g.
`cert_dir=CERT_DIR`

File: agenttest.py [hunk 3]
```diff
@@ -332,21 +332,20 @@ def send_arps(hosts):
...
     """Simple end-to-end test of FAUCET config agent"""
 
-    # Start with empty FAUCET config file
-    write_file(FAUCET.cfile, '')
-
     info('\n* Generating certificates\n')
     make_certs()
     params = dict(cert_dir=CERT_DIR, gnmi_port=GNMI_PORT, cfile=FAUCET.cfile)
     client_auth = (' -ca {cert_dir}/fakeca.crt -cert {cert_dir}/fakeclient.crt'
```
Comment (lantz) [inline agenttest.py]: Since it's Python 3, we can use an f-string (`f'...'`) here I think instead of a params dict.

File: agenttest.py [hunk 4]
```diff
@@ -358,6 +357,7 @@ def end_to_end_test():
                  ' --key {cert_dir}/fakeserver.key'
                  ' --gnmiport {gnmi_port}'
                  ' --configfile {cfile}'
+                 ' --nohup {nohup}'
                  ' --dpwait 1.0').format(**params).split()
```
Comment (lantz) [inline agenttest.py]: We can use an f-string here also I think.

File: agenttest.py [hunk 5]
```diff
@@ -438,9 +438,18 @@ def setUpClass(cls):
 
     def test_end_to_end(self):
         """Run end to end ping test"""
-        failures = end_to_end_test()
+        options = dict(nohup=0)
+        envs = dict(config_stat_reload=0)
+        failures = end_to_end_test(options, envs)
```
Comment (lantz) [inline agenttest.py]: should just be something like `end_to_end_tests(nohup=0, config_stat_reload=True)`

File: faucetagent.py [hunk 6]
```diff
@@ -65,11 +65,12 @@ def checkdeps():
...
     """Abstraction for communicating with FAUCET"""
 
-    def __init__(self,
+    def __init__(self,  # pylint: disable=too-many-arguments
                  path='/etc/faucet.yaml',
                  prometheus_port=9302,
                  timeout=120,
-                 dp_wait_fraction=0.0):
+                 dp_wait_fraction=0.0,
+                 nohup=0):
```
Comment (lantz) [inline faucetagent.py]: Probably `nohup=False`

**Comment by lantz** [comment]
p.s. I still think we should start with clearing the config file.

File: agenttest.py (hunk 4 above)
Comment (cglewis) [reply agenttest.py]: @lantz it was already as a `.format`, I was trying to minimize re-writing things for the purposes of getting this feature in. would you prefer it be rewritten for this?

**Comment by cglewis** [comment]
@lantz i can add back in clearing the config file, but the tests will fail and I'm not sure how to fix them, but we can add it back in until we figure out how to fix the failure, if you prefer?

**Comment by lantz** [comment]
Yes, I would like to figure out how to fix it.

**Comment by lantz** [comment]
Actually we've messed with this enough. I think we can just move forward for now.

**Pull request merged**